File: programs/fluid-solana-programs/tests/src/liquidity/operate_test.rs
//! Liquidity Operate Tests - Rust port of TypeScript `operate.test.ts`.
//!
//! This module contains tests for the operate function combining supply/borrow operations.

#[cfg(test)]
mod tests {
    use crate::liquidity::fixture::LiquidityFixture;
    use fluid_test_framework::helpers::MintKey;
    use fluid_test_framework::prelude::*;
    use liquidity::state::{UserBorrowConfig, UserSupplyConfig};

    const DEFAULT_SUPPLY_AMOUNT: u64 = LAMPORTS_PER_SOL;
    const U60_MAX: u128 = 1_152_921_504_606_846_975;
    const U64_MAX: u128 = u64::MAX as u128;
    const I64_MAX: u128 = i64::MAX as u128;

    fn setup_fixture_with_allowances() -> LiquidityFixture {
        let mut fixture = LiquidityFixture::new().expect("Failed to create fixture");
        fixture.setup().expect("Failed to setup fixture");

        let protocol = fixture.mock_protocol.insecure_clone();
        for mint in [MintKey::USDC, MintKey::WSOL] {
            fixture
                .set_user_allowances_default(mint, &protocol)
                .expect("Failed to set allowances for mock protocol");
        }

        let protocol_interest_free = fixture.mock_protocol_interest_free.insecure_clone();
        for mint in [MintKey::USDC, MintKey::WSOL] {
            fixture
                .set_user_allowances_default_interest_free(mint, &protocol_interest_free)
                .expect("Failed to set allowances for interest-free protocol");
        }

        fixture
    }

    fn setup_fixture_with_supply_limits(mint: MintKey) -> LiquidityFixture {
        let mut fixture = setup_fixture_with_allowances();
        let protocol_pubkey = fixture.mock_protocol.pubkey();

        let supply_config = UserSupplyConfig {
            mode: 1,
            expand_percent: 10_000,
            expand_duration: 1,
            base_withdrawal_limit: U64_MAX,
        };

        fixture
            .update_user_supply_config_with_params(mint, &protocol_pubkey, supply_config)
            .expect("Failed to update user supply config");

        fixture
    }

    /// Test: operate_RevertOperateAmountsZero
    #[test]
    fn test_operate_revert_operate_amounts_zero() {
        let mut fixture = setup_fixture_with_allowances();
        let protocol = fixture.mock_protocol.insecure_clone();
        let alice = fixture.alice.insecure_clone();

        fixture.expect_revert_any(&["OperateAmountsNearlyZero", "6001"], |f| {
            f.operate(&protocol, 0, 0, MintKey::USDC, &alice)
        });
    }

    /// Test: operate_AfterUnpaused
    #[test]
    fn test_operate_after_unpaused() {
        let mut fixture = setup_fixture_with_allowances();
        let protocol = fixture.mock_protocol.insecure_clone();
        let protocol_pubkey = protocol.pubkey();
        let alice = fixture.alice.insecure_clone();
        let liquidity_pda = fixture.get_liquidity();

        fixture
            .pause_user(MintKey::USDC, MintKey::USDC, &protocol_pubkey, 1, 1)
            .expect("Failed to pause user");

        let revert_info = fixture.expect_revert_any(&["UserPaused", "6027"], |f| {
            f.operate(
                &protocol,
                DEFAULT_SUPPLY_AMOUNT as i128,
                0,
                MintKey::USDC,
                &alice,
            )
        });

        assert!(
            revert_info.contains("UserPaused") || revert_info.has_error_code(6027),
            "Error should indicate user is paused"
        );

        let balance_before_unpaused = fixture.balance_of(&liquidity_pda, MintKey::USDC);

        fixture
            .unpause_user(MintKey::USDC, MintKey::USDC, &protocol_pubkey, 0, 0)
            .expect("Failed to unpause user");

        fixture
            .operate(
                &protocol,
                DEFAULT_SUPPLY_AMOUNT as i128,
                0,
                MintKey::USDC,
                &alice,
            )
            .expect("Operate should succeed after unpause");

        let balance_after = fixture.balance_of(&liquidity_pda, MintKey::USDC);
        assert_eq!(
            balance_after - balance_before_unpaused,
            DEFAULT_SUPPLY_AMOUNT,
            "Protocol balance should increase by supplied amount"
        );
    }

    /// Test: operate_SupplyMaxTokenAmountCapMinusOne
    #[test]
    fn test_operate_supply_max_token_amount_cap_minus_one() {
        let mint = MintKey::USDC;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol.insecure_clone();
        let alice = fixture.alice.insecure_clone();

        let supply_amount = (U60_MAX - 1) as i128;
        fixture
            .operate(&protocol, supply_amount, 0, mint, &alice)
            .expect("Supplying cap - 1 should succeed");
    }

    /// Test: operate_RevertValueOverflowTotalSupplyWithInterest
    #[test]
    fn test_operate_revert_value_overflow_total_supply_with_interest() {
        let mint = MintKey::USDC;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol.insecure_clone();
        let alice = fixture.alice.insecure_clone();

        let supply_amount = (I64_MAX - 1) as i128;
        fixture
            .operate(&protocol, supply_amount, 0, mint, &alice)
            .expect_err("Should revert with USER_MODULE_VALUE_OVERFLOW_TOTAL_SUPPLY");
    }

    /// Test: operate_WithdrawWhenAboveTotalSupplyWithInterestLimit
    #[test]
    fn test_operate_withdraw_when_above_total_supply_with_interest_limit() {
        let mint = MintKey::USDC;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol.insecure_clone();
        let alice = fixture.alice.insecure_clone();

        let max_supply = (U60_MAX - 1) as i128;
        fixture
            .operate(&protocol, max_supply, 0, mint, &alice)
            .expect("Supplying max amount should succeed");

        // Not Simulating total amounts to be > max (this would need storage manipulation in Solana)
        // For now, just test that withdraw works when at limit
        let withdraw_amount = (U60_MAX / 2) as i128;
        fixture
            .operate(&protocol, -withdraw_amount, 0, mint, &alice)
            .expect("Withdraw at limit should succeed");
    }

    /// Test: operate_RevertValueOverflowTotalSupplyInterestFree
    #[test]
    fn test_operate_revert_value_overflow_total_supply_interest_free() {
        let mint = MintKey::USDC;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol_interest_free.insecure_clone();
        let alice = fixture.alice.insecure_clone();

        let initial_supply: i128 = 1_000_000;
        fixture
            .operate(&protocol, initial_supply, 0, mint, &alice)
            .expect("Initial supply should succeed");

        let overflow_amount = (I64_MAX as i128) - initial_supply;
        fixture.expect_revert_with("ValueOverflowTotalSupply", |f| {
            f.operate(&protocol, overflow_amount, 0, mint, &alice)
        });
    }

    /// Test: operate_RevertValueOverflowTotalBorrowWithInterest
    #[test]
    fn test_operate_revert_value_overflow_total_borrow_with_interest() {
        let mint = MintKey::WSOL;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol.insecure_clone();
        let protocol_pubkey = protocol.pubkey();
        let alice = fixture.alice.insecure_clone();

        let borrow_config = UserBorrowConfig {
            mode: 1,
            expand_percent: 0,
            expand_duration: 1,
            base_debt_ceiling: U60_MAX + 1,
            max_debt_ceiling: U60_MAX + 1,
        };
        fixture
            .update_user_borrow_config_with_params(mint, &protocol_pubkey, borrow_config)
            .expect("Failed to update borrow config");

        let max_amount = U60_MAX as i128;
        fixture
            .operate(&protocol, max_amount - 1, 0, mint, &alice)
            .expect("Supplying near max should succeed");

        let borrow_half = (U60_MAX / 2) as i128;
        fixture
            .operate(&protocol, 0, borrow_half, mint, &alice)
            .expect("Borrow up to half should work");

        fixture.expect_revert_with("ValueOverflowTotalBorrow", |f| {
            f.operate(&protocol, 0, borrow_half + 2, mint, &alice)
        });
    }

    /// Test: operate_PaybackWhenAboveTotalBorrowWithInterestLimit
    #[test]
    fn test_operate_payback_when_above_total_borrow_with_interest_limit() {
        let mint = MintKey::WSOL;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol.insecure_clone();
        let protocol_pubkey = protocol.pubkey();
        let alice = fixture.alice.insecure_clone();

        let borrow_config = UserBorrowConfig {
            mode: 1,
            expand_percent: 0,
            expand_duration: 1,
            base_debt_ceiling: U60_MAX + 1,
            max_debt_ceiling: U60_MAX + 1,
        };
        fixture
            .update_user_borrow_config_with_params(mint, &protocol_pubkey, borrow_config)
            .expect("Failed to update borrow config");

        let buffer: u128 = 10_000_000;
        let max_amount = (U60_MAX - buffer) as i128;

        fixture
            .operate(&protocol, max_amount, 0, mint, &alice)
            .expect("Supplying max amount should succeed");

        let borrow_amount = (U60_MAX - buffer) as i128;
        fixture
            .operate(&protocol, 0, borrow_amount, mint, &alice)
            .expect("Borrow near cap should succeed");

        let payback_amount = buffer as i128;
        fixture
            .operate(&protocol, 0, -payback_amount, mint, &alice)
            .expect("Payback should succeed when above total borrow limit");
    }

    /// Test: operate_RevertValueOverflowTotalBorrowInterestFree
    #[test]
    fn test_operate_revert_value_overflow_total_borrow_interest_free() {
        let mint = MintKey::WSOL;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol_interest_free.insecure_clone();
        let protocol_pubkey = protocol.pubkey();
        let alice = fixture.alice.insecure_clone();

        let borrow_config = UserBorrowConfig {
            mode: 0,
            expand_percent: 0,
            expand_duration: 1,
            base_debt_ceiling: U60_MAX + 1,
            max_debt_ceiling: U60_MAX + 1,
        };
        fixture
            .update_user_borrow_config_with_params(mint, &protocol_pubkey, borrow_config)
            .expect("Failed to update borrow config");

        let buffer: u128 = 10_000_000;
        let max_amount = (U60_MAX - buffer) as i128;

        fixture
            .operate(&protocol, max_amount, 0, mint, &alice)
            .expect("Supply should succeed for interest-free protocol");

        fixture
            .operate(&protocol, 0, max_amount, mint, &alice)
            .expect("Borrow up to max should succeed");

        fixture.expect_revert_any(&["ValueOverflowTotalBorrow", "6034"], |f| {
            f.operate(&protocol, 0, (buffer + 1) as i128, mint, &alice)
        });
    }

    /// Test: operate_OperateAmountWorksForWithdraw
    #[test]
    fn test_operate_amount_works_for_withdraw() {
        let mint = MintKey::USDC;
        let mut fixture = setup_fixture_with_supply_limits(mint);
        let protocol = fixture.mock_protocol.insecure_clone();
        let alice = fixture.alice.insecure_clone();

        let initial_supply = (U60_MAX - 1_000) as i128;
        fixture
            .operate(&protocol, initial_supply, 0, mint, &alice)
            .expect("Supplying large amount should succeed");

        fixture
            .operate(&protocol, -10, 0, mint, &alice)
            .expect("Withdraw 10 should succeed");

        fixture
            .operate(&protocol, -100, 0, mint, &alice)
            .expect("Withdraw 100 should succeed");
    }
}

File: programs/fluid-solana-programs/__tests__/liquidity/operate.test.ts
import { expect } from "chai";
import { BN } from "@coral-xyz/anchor";
import { LAMPORTS_PER_SOL } from "@solana/web3.js";

import {
  UpdateUserSupplyConfigParams,
  UpdateUserBorrowConfigParams,
} from "../../ts-sdk/liquidity/types";
import { MintKeys, mint as MintInfo } from "../../ts-sdk/mint";
import { LiquidityBaseSetup } from "../../test-utils/typescript/liquidity/setup";

const DEFAULT_SUPPLY_AMOUNT = new BN(1 * LAMPORTS_PER_SOL);

describe("LiquidityUserModuleOperateTests", () => {
  let setup: LiquidityBaseSetup;

  afterEach(async () => {
    setup.logComputeBudget();
  });

  beforeEach(async () => {
    process.env.TEST_MODE_JEST = "true";

    setup = new LiquidityBaseSetup();
    await setup.setup();
    await setup._setUserAllowancesDefault(MintKeys.USDC, setup.mockProtocol);
  });

  it("Should test: operate_RevertOperateAmountsZero", async () => {
    expect(
      await setup.expectRevert(
        "USER_MODULE_OPERATE_AMOUNTS_ZERO",
        setup.operate.bind(setup),
        setup.mockProtocol,
        new BN(0), // supply amount
        new BN(0), // borrow amount
        MintKeys.USDC,
        setup.alice
      )
    ).to.be.true;
  });

  it("Should test: operate_AfterUnpaused", async () => {
    const balanceBefore = await setup.balanceOf(
      setup.mockProtocol.publicKey,
      MintInfo.getMint(MintKeys.USDC)
    );

    // Pause user
    await setup.pauseUser(
      MintKeys.USDC,
      MintKeys.USDC,
      setup.mockProtocol.publicKey,
      1,
      1
    );

    expect(
      await setup.expectRevert(
        "USER_MODULE_USER_PAUSED",
        setup.operate.bind(setup),
        setup.mockProtocol,
        DEFAULT_SUPPLY_AMOUNT,
        new BN(0),
        MintKeys.USDC,
        setup.alice
      )
    ).to.be.true;

    // Now unpause and execute again
    await setup.unpauseUser(
      MintKeys.USDC,
      MintKeys.USDC,
      setup.mockProtocol.publicKey,
      0,
      0
    );

    await setup.operate(
      setup.mockProtocol,
      DEFAULT_SUPPLY_AMOUNT,
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    const balanceAfter = await setup.balanceOf(
      setup.mockProtocol.publicKey,
      MintInfo.getMint(MintKeys.USDC)
    );

    expect(balanceAfter.sub(balanceBefore).toString()).to.equal(
      DEFAULT_SUPPLY_AMOUNT.toString()
    );
  });
});

describe("LiquidityUserModuleOperateTotalAmountsOverflowTests", () => {
  let setup: LiquidityBaseSetup;
  const u64MAX = new BN("18446744073709551615");
  const u60MAX = new BN("1152921504606846975");
  const i64MAX = new BN("9223372036854775807");

  afterEach(async () => {
    setup.logComputeBudget();
  });

  beforeEach(async () => {
    process.env.TEST_MODE_JEST = "true";

    setup = new LiquidityBaseSetup();
    await setup.setup();

    // manually set user allowances to 1 for all mints
    // Set withdraw config with actual limits
    const supplyConfig: UpdateUserSupplyConfigParams = {
      mint: MintKeys.USDC,
      newSupplyConfig: {
        user: setup.mockProtocol.publicKey,
        token: MintInfo.getMint(MintKeys.USDC),
        mode: 1, // with interest
        expandPercent: new BN(1e4), // 100%
        expandDuration: new BN(1), // every second
        baseWithdrawalLimit: u64MAX,
      },
    };

    await setup.updateUserSupplyConfig([supplyConfig]);
  });

  it("Should test: supply_MAX_TOKEN_AMOUNT_CAP_minus_1", async () => {
    await setup.operate(
      setup.mockProtocol,
      u60MAX.sub(new BN(1)),
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );
  });

  it("Should test: operate_RevertValueOverflowTotalSupplyWithInterest", async () => {
    // Try to supply above max and expect revert
    expect(
      await setup.expectRevert(
        "USER_MODULE_VALUE_OVERFLOW_TOTAL_SUPPLY",
        setup.operate.bind(setup),
        setup.mockProtocol,
        i64MAX.sub(new BN(1)), // u64::MAX
        new BN(0),
        MintKeys.USDC,
        setup.alice
      )
    ).to.be.true;
  });

  it("Should test: operate_WithdrawWhenAboveTotalSupplyWithInterestLimit", async () => {
    const maxSupply = u60MAX.sub(new BN(1));

    // Supply max amount first
    await setup.operate(
      setup.mockProtocol,
      maxSupply,
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    // Not Simulating total amounts to be > max (this would need storage manipulation in Solana)
    // For now, just test that withdraw works when at limit
    const withdrawAmount = u60MAX.div(new BN(2));

    // Expect withdraw to work
    await setup.operate(
      setup.mockProtocol,
      withdrawAmount.neg(), // negative for withdraw
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );
  });

  it("Should test: operate_RevertValueOverflowTotalSupplyInterestFree", async () => {
    const initialSupply = new BN(1e6);

    // Supply max amount first with interest-free protocol
    await setup.operate(
      setup.mockProtocolInterestFree,
      initialSupply,
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    // Try to supply above max and expect revert
    expect(
      await setup.expectRevert(
        "USER_MODULE_VALUE_OVERFLOW_TOTAL_SUPPLY",
        setup.operate.bind(setup),
        setup.mockProtocolInterestFree,
        i64MAX.sub(initialSupply),
        new BN(0),
        MintKeys.USDC,
        setup.alice
      )
    ).to.be.true;
  });

  it("Should test: operate_RevertValueOverflowTotalBorrowWithInterest", async () => {
    // Mint huge amount for testing
    // Note: In Solana, you'd need to mint to the test account

    // Increase borrow limits
    const borrowConfig: UpdateUserBorrowConfigParams = {
      mint: MintKeys.USDC,
      newBorrowConfig: {
        user: setup.mockProtocol.publicKey,
        token: MintInfo.getMint(MintKeys.USDC),
        mode: 1, // with interest
        expandPercent: new BN(0),
        expandDuration: new BN(1),
        baseDebtCeiling: u60MAX.add(new BN(1)),
        maxDebtCeiling: u60MAX.add(new BN(1)),
      },
    };

    await setup.updateUserBorrowConfig([borrowConfig]);

    const maxAmount = u60MAX;

    // Supply max
    await setup.operate(
      setup.mockProtocol,
      maxAmount.sub(new BN(1)), // 1152921504606846974
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    // Borrow max.div(new BN(2))
    await setup.operate(
      setup.mockProtocol,
      new BN(0),
      maxAmount.div(new BN(2)), // 576460752303423487
      MintKeys.USDC,
      setup.alice
    );

    // Try to borrow above max and expect revert
    expect(
      await setup.expectRevert(
        "USER_MODULE_VALUE_OVERFLOW_TOTAL_BORROW",
        setup.operate.bind(setup),
        setup.mockProtocol,
        new BN(0),
        maxAmount.div(new BN(2)).add(new BN(2)), // This will make total borrow to be just above current_borrow_limit, but below max_cap_amount
        MintKeys.USDC,
        setup.alice
      )
    ).to.be.true;
  });

  it("Should test: operate_PaybackWhenAboveTotalBorrowWithInterestLimit", async () => {
    // Similar to above but test payback functionality
    const borrowConfig: UpdateUserBorrowConfigParams = {
      mint: MintKeys.USDC,
      newBorrowConfig: {
        user: setup.mockProtocol.publicKey,
        token: MintInfo.getMint(MintKeys.USDC),
        mode: 1,
        expandPercent: new BN(0),
        expandDuration: new BN(1),
        baseDebtCeiling: u60MAX.add(new BN(1)),
        maxDebtCeiling: u60MAX.add(new BN(1)),
      },
    };

    await setup.updateUserBorrowConfig([borrowConfig]);

    const maxAmount = u60MAX.sub(new BN(10e6));

    // Supply and borrow max
    await setup.operate(
      setup.mockProtocol,
      maxAmount,
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    await setup.operate(
      setup.mockProtocol,
      new BN(0),
      u60MAX.sub(new BN(10e6)), // some buffer for rounding
      MintKeys.USDC,
      setup.alice
    );

    // Expect payback to work
    const paybackAmount = new BN(10e6);

    await setup.operate(
      setup.mockProtocol,
      new BN(0),
      paybackAmount.neg(), // negative for payback
      MintKeys.USDC,
      setup.alice
    );
  });

  it("Should test: operate_RevertValueOverflowTotalBorrowInterestFree", async () => {
    const borrowConfig: UpdateUserBorrowConfigParams = {
      mint: MintKeys.USDC,
      newBorrowConfig: {
        user: setup.mockProtocolInterestFree.publicKey,
        token: MintInfo.getMint(MintKeys.USDC),
        mode: 0, // interest free
        expandPercent: new BN(0),
        expandDuration: new BN(1),
        baseDebtCeiling: u60MAX.add(new BN(1)),
        maxDebtCeiling: u60MAX.add(new BN(1)),
      },
    };

    await setup.updateUserBorrowConfig([borrowConfig]);

    const maxAmount = u60MAX.sub(new BN(10e6));

    // Supply max
    await setup.operate(
      setup.mockProtocolInterestFree,
      maxAmount,
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    // Borrow max
    await setup.operate(
      setup.mockProtocolInterestFree,
      new BN(0),
      maxAmount,
      MintKeys.USDC,
      setup.alice
    );

    // Try to borrow above max and expect revert
    expect(
      await setup.expectRevert(
        "USER_MODULE_VALUE_OVERFLOW_TOTAL_BORROW",
        setup.operate.bind(setup),
        setup.mockProtocolInterestFree,
        new BN(0),
        new BN(10e6).add(new BN(1)),
        MintKeys.USDC,
        setup.alice
      )
    ).to.be.true;
  });
});

describe("LiquidityUserModuleOperateAmountInsufficientTests", () => {
  let setup: LiquidityBaseSetup;
  const u64MAX = new BN("18446744073709551615");
  const u60MAX = new BN("1152921504606846975");

  afterEach(async () => {
    setup.logComputeBudget();
  });

  beforeEach(async () => {
    process.env.TEST_MODE_JEST = "true";

    setup = new LiquidityBaseSetup();
    await setup.setup();

    // manually set user allowances to 1 for all mints
    // Set withdraw config with actual limits
    const supplyConfig: UpdateUserSupplyConfigParams = {
      mint: MintKeys.USDC,
      newSupplyConfig: {
        user: setup.mockProtocol.publicKey,
        token: MintInfo.getMint(MintKeys.USDC),
        mode: 1, // with interest
        expandPercent: new BN(1e4), // 100%
        expandDuration: new BN(1), // every second
        baseWithdrawalLimit: u64MAX,
      },
    };

    await setup.updateUserSupplyConfig([supplyConfig]);
  });

  // Test deprecated, as we are not using bigNumber anymore
  it("Should test: operate_RevertOperateAmountInsufficientWhenUserSupply", async () => {
    // Supply large amount first
    // await setup.operate(
    //   setup.mockProtocol,
    //   u64MAX.sub(new BN(1e3)), // 1e28
    //   new BN(0),
    //   MintKeys.USDC,
    //   setup.alice
    // );
    // // Try to supply insufficient amount and expect revert
    // // 1e28 - big math precision ~1e17 -> 1e11 will be too small
    // expect(
    //   await setup.expectRevert(
    //     "USER_MODULE_OPERATE_AMOUNT_INSUFFICIENT",
    //     setup.operate.bind(setup),
    //     setup.mockProtocol,
    //     new BN(1e2), // 1e11
    //     new BN(0),
    //     MintKeys.USDC,
    //     setup.alice
    //   )
    // ).to.be.true;
    // Not able to test it, as there is restriction on totalSupply check on token transfers
    // // Expect deposit from other user (protocol) to work
    // await setup.operate(
    //   setup.mockProtocolWithInterest,
    //   new BN(1e2), // 1e6
    //   new BN(0),
    //   MintKeys.USDC,
    //   setup.bob
    // );
  });

  it("Should test: operate_OperateAmountWorksForWithdraw", async () => {
    // Supply large amount first
    await setup.operate(
      setup.mockProtocol,
      u60MAX.sub(new BN(1e3)), // 1e28
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    await setup.operate(
      setup.mockProtocol,
      new BN(-10), // withdraw 1
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );

    await setup.operate(
      setup.mockProtocol,
      new BN(-100), // withdraw 100
      new BN(0),
      MintKeys.USDC,
      setup.alice
    );
  });

  // it("Should test: operate_RevertOperateAmountInsufficientWhenTotalSupply", async () => {
  //   // Supply large amount first
  //   await setup.operate(
  //     setup.mockProtocol,
  //     u64MAX.sub(new BN(1e3)), // 1e28
  //     new BN(0),
  //     MintKeys.USDC,
  //     setup.alice
  //   );

  //   // Expect deposit to revert if operate amount is too small
  //   expect(
  //     await setup.expectRevert(
  //       "USER_MODULE_OPERATE_AMOUNT_INSUFFICIENT",
  //       setup.operate.bind(setup),
  //       setup.mockProtocolWithInterest,
  //       new BN("100000000000"), // 1e11
  //       new BN(0),
  //       MintKeys.USDC,
  //       setup.alice
  //     )
  //   ).to.be.true;

  //   // Bigger amount will work
  //   await setup.operate(
  //     setup.mockProtocolWithInterest,
  //     new BN("1000000000000"), // 1e12
  //     new BN(0),
  //     MintKeys.USDC,
  //     setup.alice
  //   );
  // });

  // it("Should test: operate_OperateAmountWorksForBorrow", async () => {
  //   // Increase borrow limits
  //   const borrowConfigs: BORROW_CONFIG[] = [
  //     {
  //       mint: MintKeys.USDC,
  //       newBorrowConfig: {
  //         user: setup.mockProtocolWithInterest.publicKey,
  //         token: MintInfo.getMint(MintKeys.USDC),
  //         mode: 1,
  //         expandPercent: new BN(0),
  //         expandDuration: new BN(1),
  //         baseDebtCeiling: new BN("1000000000000000000000000000000000"), // 1e32
  //         maxDebtCeiling: new BN("10000000000000000000000000000000000"), // 1e33
  //       },
  //     },
  //     {
  //       mint: MintKeys.USDC,
  //       newBorrowConfig: {
  //         user: setup.mockProtocol.publicKey,
  //         token: MintInfo.getMint(MintKeys.USDC),
  //         mode: 1,
  //         expandPercent: new BN(0),
  //         expandDuration: new BN(1),
  //         baseDebtCeiling: new BN("1000000000000000000000000000000000"), // 1e32
  //         maxDebtCeiling: new BN("10000000000000000000000000000000000"), // 1e33
  //       },
  //     },
  //   ];
  //   await setup.updateUserBorrowConfig(borrowConfigs);

  //   // Supply large amount
  //   await setup.operate(
  //     setup.mockProtocol,
  //     new BN("1000000000000000000000000000000"), // 1e30
  //     new BN(0),
  //     MintKeys.USDC,
  //     setup.alice
  //   );

  //   // Borrow
  //   await setup.operate(
  //     setup.mockProtocol,
  //     new BN(0),
  //     new BN("100000000000000000000000000000"), // 1e28
  //     MintKeys.USDC,
  //     setup.alice
  //   );

  //   // Borrow is always rounded up so it ALWAYS triggers a change in user borrow amount even if operate amount is only 1
  //   await setup.operate(
  //     setup.mockProtocol,
  //     new BN(0),
  //     new BN(1),
  //     MintKeys.USDC,
  //     setup.alice
  //   );

  //   // Expect borrow from other user (protocol) to work
  //   await setup.operate(
  //     setup.mockProtocolWithInterest,
  //     new BN(0),
  //     new BN("100000000000"), // 1e11
  //     MintKeys.USDC,
  //     setup.alice
  //   );
  // });
});
